### PR TITLE
RegEx improvements

### PR DESCRIPTION
**Pull request by StartAutomating**
Structuring Regular Expressions used within arm-ttk

### CODE REVIEW

**Comment by bmoore-msft** [comment]
Let's expand on the docs and make them more scenario focused - so when would I use one approach vs. the other and show an example of consuming in a test...

**Comment by StartAutomating** [comment]
We would like some samples of simple patterns / generators inline with /Regex/README.md

File: arm-ttk/Regex/README.md [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+*.regex.txt pattern files
+*.regex.ps1 pattern generators
+
+These regular expressions can be used within arm-ttk by calling a PowerShell ScriptBlock that matches the name of the directory/file.
+
+### Using Patterns
+
+
+.regex.txt files will contain a single Regular Expression.
+
```
Comment (bmoore-msft) [inline arm-ttk/Regex/README.md:21]: Let's add an inline example for the content of the txt file in a code block

File: arm-ttk/Regex/README.md [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+This expression can be used anywhere within arm-ttk.  
+
+For instance, using the pattern ARM_Expression, we can find all expressions within the text of a template:
```
Comment (bmoore-msft) [inline arm-ttk/Regex/README.md:24]: add a step to talk about the hierarchy and how the name is derived, i.e. where does "ARM_Expression" come from

File: arm-ttk/Regex/README.md [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+~~~PowerShell
+$templateText | & ${?<ARM_Expression>}
```
Comment (bmoore-msft) [inline arm-ttk/Regex/README.md:27]: add - this will find all the "matches" for the expression in templateText

File: arm-ttk/Regex/README.md [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+~~~
+
+
+### Using Generators
+
+.regex.ps1 files will contain Regular Expression generators.
+
```
Comment (bmoore-msft) [inline arm-ttk/Regex/README.md:34]: same as above lets put the the code inline

And we can also use a simpler example (doesn't have to be real) to highlight the use of a parameter to make the regex more flexible, e.g. whitespace+param+whitespace

File: arm-ttk/Regex/README.md [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+These are PowerShell scripts that produce a regular expression based off of input.
+
+For instance, we can find all of the references to the VMSize parameter.
+
+~~~PowerShell
+$templateText | & ${?<ARM_Parameter>} -Parameter 'VMSize'
+~~~
+
+Generators need not require parameters.  To find all parameters referenced throughout the template, use:
```
Comment (bmoore-msft) [inline arm-ttk/Regex/README.md:43]: you could use a generator to find all resources of a given type and return properties for that resource

File: arm-ttk/Regex/ARM/Expression.regex.txt [hunk 6]
```diff
@@ -0,0 +1,14 @@
+﻿# Matches an Expression in an Azure Resource Manager Template
+\s{0,}                                                     # optional whitespace
```
Comment (bmoore-msft) [inline arm-ttk/Regex/ARM/Expression.regex.txt:2]: can  we add some more detail up top about the depth and balancing groups piece (i.e. why is it relevant to the scenario)

File: arm-ttk/Regex/ARM/List_Function.regex.txt [hunk 7]
```diff
@@ -0,0 +1,9 @@
+﻿(?>        # we don't want to flag a UDF that might be called "myListOfIps" so we need to check the char preceeding list*()
```
Comment (bmoore-msft) [inline arm-ttk/Regex/ARM/List_Function.regex.txt:1]: Or's should be more explicit in the comments

File: arm-ttk/Writing_TTK_Tests.md [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+            # At this point, $foundQuote could contain either a double quote or a bracket
+            # It it was a double quote, the parameter reference is not in an expression.
+            if ($foundQuote.Value -eq '"') { 
+                # so we produce an error.
+                Write-Error -Message "Parameter reference is not contained within an expression: $($Parameter.Name)" -ErrorId Parameters.Must.Be.Referenced.In.Expression -TargetObject $parameter
+            }
+        }
+    }
+}
+~~~
```
Comment (bmoore-msft) [inline arm-ttk/Writing_TTK_Tests.md:85]: Add a section or another file for how to supply unit tests

**Review by bmoore-msft — CHANGES_REQUESTED** [review]
Remember to pull first since we made a couple changes during the review

File: arm-ttk/Regex/README.md (hunk 4 above)
Comment (StartAutomating) [reply arm-ttk/Regex/README.md:57]: Example / "Writing Generator" section added

File: arm-ttk/Regex/README.md (hunk 5 above)
Comment (StartAutomating) [reply arm-ttk/Regex/README.md:66]: We have determined we want a simpler example, using ARM_Parameter.

**Comment by StartAutomating** [comment]
Upon discussion, we would also like to rename the following expressions:  ARM_Parameter -> ARM_Template_Parameter and ARM_Variable - > ARM_Template_Variable.

Any future expression whose name would be easily confused without clarifying should be named ARM_Template_*

File: arm-ttk/Regex/ARM/List_Function.regex.txt [hunk 9]
```diff
@@ -1,9 +1,20 @@
...
+    \[     # A bracket (the function could be the first thing in the template expression)
+    |      # OR
+    \(     # An open parenthesis (the function may be contained in another function)
+    |      # OR
+    ,      # A comma (the function may be contained in another function and not the first argument in that function)
+    )      # and the (?>  ) syntax says this is not included in the match because we need to check for expressions explicitly below    
+    \s{0,} # We also don't need to include any preceeding whitespace in the match itself
+)
+(?<ListFunctionName>  # We want to capture the list function name
+list\w{1,}            # which will be list followed by any number of word characters
```
Comment (bmoore-msft) [inline arm-ttk/Regex/ARM/List_Function.regex.txt:15]: should we just do this with the TLE generator?

**Pull request merged**